From 45a21746797a934f2cac41d34149332aa122ea12 Mon Sep 17 00:00:00 2001
From: LR-Vensuki <260785366+LR-Vensuki@users.noreply.github.com>
Date: Thu, 8 Oct 2026 06:09:05 +0300
Subject: [PATCH 11/42] armv7 iOS 6: CoreText font manager and color spaces
 through iOS 4 calls, kevent64 polling, NativeLayerCA colors without iOS 13
 APIs, H.264-only hardware decode check, nsWindow touch radius

---
 .../platforms/apple/AppleDecoderModule.cpp    | 10 ++++++++
 gfx/2d/MacIOSurface.cpp                       |  3 +++
 gfx/2d/NativeFontResourceMac.cpp              |  3 +++
 gfx/layers/NativeLayerCA.mm                   | 23 ++++++++++++++++++-
 gfx/thebes/CoreTextFontList.cpp               |  3 +++
 gfx/thebes/gfxPlatformMac.cpp                 |  4 ++++
 ipc/chromium/src/base/message_pump_kqueue.cc  |  9 ++++++++
 toolkit/components/processtools/ProcInfo.mm   |  2 +-
 widget/uikit/nsWindow.mm                      | 12 +++++++++-
 9 files changed, 66 insertions(+), 3 deletions(-)

diff --git a/dom/media/platforms/apple/AppleDecoderModule.cpp b/dom/media/platforms/apple/AppleDecoderModule.cpp
index 27c25b050d..43268e578e 100644
--- a/dom/media/platforms/apple/AppleDecoderModule.cpp
+++ b/dom/media/platforms/apple/AppleDecoderModule.cpp
@@ -38,7 +38,11 @@ static inline CMVideoCodecType GetCMVideoCodecType(const MediaCodec& aCodec) {
     case MediaCodec::AV1:
       return kCMVideoCodecType_AV1;
     case MediaCodec::VP9:
+#if defined(MOZ_IOS_LEGACY)  // not in the iOS 6.1 SDK
+      return static_cast<CMVideoCodecType>('vp09');
+#else
       return kCMVideoCodecType_VP9;
+#endif
     case MediaCodec::HEVC:
       return kCMVideoCodecType_HEVC;
     default:
@@ -238,9 +242,15 @@ bool AppleDecoderModule::CanCreateHWDecoder(const MediaCodec& aCodec) {
     return false;
   }
 
+#if defined(MOZ_IOS_LEGACY)
+  // VTIsHardwareDecodeSupported is iOS 11. A4-A6 decode H.264 in hardware
+  // and nothing else.
+  return aCodec == MediaCodec::H264;
+#else
   if (!VTIsHardwareDecodeSupported(GetCMVideoCodecType(aCodec))) {
     return false;
   }
+#endif
 
   // H264 hardware decoding has been supported since macOS 10.6 on most Intel
   // GPUs (Sandy Bridge and later, 2011). If VTIsHardwareDecodeSupported is
diff --git a/gfx/2d/MacIOSurface.cpp b/gfx/2d/MacIOSurface.cpp
index fdc77a2790..3ba74e97a6 100644
--- a/gfx/2d/MacIOSurface.cpp
+++ b/gfx/2d/MacIOSurface.cpp
@@ -20,6 +20,9 @@
 #include "mozilla/RefPtr.h"
 #include "mozilla/gfx/Logging.h"
 #include "mozilla/StaticPrefs_gfx.h"
+#if defined(MOZ_IOS_LEGACY)
+#  include <legacyfox/ColorSpaceNames.h>  // names from newer SDKs
+#endif
 
 using namespace mozilla;
 
diff --git a/gfx/2d/NativeFontResourceMac.cpp b/gfx/2d/NativeFontResourceMac.cpp
index 5f97432e84..50ae52b1f4 100644
--- a/gfx/2d/NativeFontResourceMac.cpp
+++ b/gfx/2d/NativeFontResourceMac.cpp
@@ -15,6 +15,9 @@
 #endif
 
 #include "nsIMemoryReporter.h"
+#if defined(MOZ_IOS_LEGACY)
+#  include <legacyfox/CoreTextCompat.h>  // font manager calls from iOS 13
+#endif
 
 namespace mozilla {
 namespace gfx {
diff --git a/gfx/layers/NativeLayerCA.mm b/gfx/layers/NativeLayerCA.mm
index 94efd9b6c3..41c09dd741 100644
--- a/gfx/layers/NativeLayerCA.mm
+++ b/gfx/layers/NativeLayerCA.mm
@@ -10,6 +10,9 @@
 #  import <OpenGL/gl.h>
 #endif
 #import <AVFoundation/AVFoundation.h>
+#if defined(MOZ_IOS_LEGACY)
+#  import <legacyfox/AVSampleBufferDisplayLayer.h>
+#endif
 #import <QuartzCore/QuartzCore.h>
 
 #include <algorithm>
@@ -414,7 +417,7 @@ void NativeLayerRootCA::CommitRepresentation(
   // the sublayers array - layers which are completely clipped out will return
   // null from UndelyingCALayer.
   AutoCATransaction transaction;
-  NSMutableArray<CALayer*>* sublayers =
+  NSMutableArray* sublayers =
       [NSMutableArray arrayWithCapacity:aSublayers.Length()];
   bool mustRebuild = updateRequired == UpdateType::All;
   for (const auto& layer : aSublayers) {
@@ -497,9 +500,18 @@ void NativeLayerRootCA::SetWindowIsFullscreen(bool aFullscreen) {
 }
 
 /* static */ bool IsCGColorOpaqueBlack(CGColorRef aColor) {
+#if defined(MOZ_IOS_LEGACY)  // no constant colors on iOS before 13
+  if (aColor && CGColorGetNumberOfComponents(aColor) == 4) {
+    const CGFloat* c = CGColorGetComponents(aColor);
+    if (c[0] == 0 && c[1] == 0 && c[2] == 0 && c[3] == 1) {
+      return true;
+    }
+  }
+#else
   if (CGColorEqualToColor(aColor, CGColorGetConstantColor(kCGColorBlack))) {
     return true;
   }
+#endif
   size_t componentCount = CGColorGetNumberOfComponents(aColor);
   if (componentCount == 0) {
     // This will happen if aColor is kCGColorClear. It's not opaque black.
@@ -809,11 +821,20 @@ NativeLayerCA::NativeLayerCA(bool aIsOpaque)
 }
 
 CGColorRef CGColorCreateForDeviceColor(const gfx::DeviceColor& aColor) {
+#if defined(MOZ_IOS_LEGACY)
+  // iOS 4-7 have only the device RGB space for this.
+  CGColorSpaceRef space = CGColorSpaceCreateDeviceRGB();
+  const CGFloat components[] = {aColor.r, aColor.g, aColor.b, aColor.a};
+  CGColorRef color = CGColorCreate(space, components);
+  CGColorSpaceRelease(space);
+  return color;
+#else
   if (StaticPrefs::gfx_color_management_native_srgb()) {
     return CGColorCreateSRGB(aColor.r, aColor.g, aColor.b, aColor.a);
   }
 
   return CGColorCreateGenericRGB(aColor.r, aColor.g, aColor.b, aColor.a);
+#endif
 }
 
 NativeLayerCA::NativeLayerCA(gfx::DeviceColor aColor)
diff --git a/gfx/thebes/CoreTextFontList.cpp b/gfx/thebes/CoreTextFontList.cpp
index 0efbed8d2a..98454ae471 100644
--- a/gfx/thebes/CoreTextFontList.cpp
+++ b/gfx/thebes/CoreTextFontList.cpp
@@ -30,6 +30,9 @@
 #include "nsIDirectoryEnumerator.h"
 #include "nsServiceManagerUtils.h"
 #include "SharedFontList-impl.h"
+#if defined(MOZ_IOS_LEGACY)
+#  include <legacyfox/CoreTextCompat.h>  // font manager calls from iOS 13
+#endif
 
 using namespace mozilla;
 using namespace mozilla::gfx;
diff --git a/gfx/thebes/gfxPlatformMac.cpp b/gfx/thebes/gfxPlatformMac.cpp
index 12d8567ec8..04dc47c7a5 100644
--- a/gfx/thebes/gfxPlatformMac.cpp
+++ b/gfx/thebes/gfxPlatformMac.cpp
@@ -1058,7 +1058,11 @@ nsTArray<uint8_t> gfxPlatformMac::GetPlatformCMSOutputProfileData() {
     return nsTArray<uint8_t>();
   }
 
+#if defined(MOZ_IOS_LEGACY)
+  CFDataRef iccp = nullptr;  // CGColorSpaceCopyICCData is iOS 10; no profile
+#else
   CFDataRef iccp = ::CGColorSpaceCopyICCData(cspace);
+#endif
 
   ::CFRelease(cspace);
 
diff --git a/ipc/chromium/src/base/message_pump_kqueue.cc b/ipc/chromium/src/base/message_pump_kqueue.cc
index 46c1d5363a..6dd2c27ce0 100644
--- a/ipc/chromium/src/base/message_pump_kqueue.cc
+++ b/ipc/chromium/src/base/message_pump_kqueue.cc
@@ -18,6 +18,10 @@
 
 namespace base {
 
+#if defined(MOZ_IOS_LEGACY) && !defined(KEVENT_FLAG_IMMEDIATE)
+#  define KEVENT_FLAG_IMMEDIATE 0x001
+#endif
+
 namespace {
 
 // On iOS, the normal `kevent64` method is blocked by the content process
@@ -26,6 +30,11 @@ static int platform_kevent64(int fd, const kevent64_s* changelist, int nchanges,
                              kevent64_s* eventlist, int nevents, int flags) {
 #if defined(XP_IOS) && !defined(MOZ_IOS_LEGACY)
   return be_kevent64(fd, changelist, nchanges, eventlist, nevents, flags);
+#elif defined(MOZ_IOS_LEGACY)
+  // The iOS 6 SDK has no KEVENT_FLAG_IMMEDIATE: a zero timeout polls the same.
+  static const struct timespec kZero = {0, 0};
+  return kevent64(fd, changelist, nchanges, eventlist, nevents, 0,
+                  flags ? &kZero : nullptr);
 #else
   return kevent64(fd, changelist, nchanges, eventlist, nevents, flags, nullptr);
 #endif
diff --git a/toolkit/components/processtools/ProcInfo.mm b/toolkit/components/processtools/ProcInfo.mm
index 95d329b0aa..f19316d57b 100644
--- a/toolkit/components/processtools/ProcInfo.mm
+++ b/toolkit/components/processtools/ProcInfo.mm
@@ -39,7 +39,7 @@ static bool LegacyTaskInfo(task_t aTask, uint64_t* aCpuNs, uint64_t* aResident)
                   PR_NSEC_PER_SEC +
               (uint64_t(info.user_time.microseconds) +
                info.system_time.microseconds) *
-                  PR_NSEC_PER_USEC;
+                  1000;  // ns per us
   }
   if (aResident) {
     *aResident = info.resident_size;
diff --git a/widget/uikit/nsWindow.mm b/widget/uikit/nsWindow.mm
index 01907f7f11..07e1f5a0e6 100644
--- a/widget/uikit/nsWindow.mm
+++ b/widget/uikit/nsWindow.mm
@@ -151,7 +151,9 @@ class nsAutoRetainUIKitObject {
 - (nullable id)accessibilityElementAtIndex:(NSInteger)index;
 - (NSInteger)indexOfAccessibilityElement:(id)element;
 - (NSArray* _Nullable)accessibilityElements;
+#  if !defined(MOZ_IOS_LEGACY)  // UIAccessibilityContainerType is iOS 11
 - (UIAccessibilityContainerType)accessibilityContainerType;
+#  endif
 #endif
 
 @end
@@ -271,9 +273,15 @@ class nsAutoRetainUIKitObject {
   for (UITouch* touch in aTouches) {
     LayoutDeviceIntPoint loc = UIKitPointsToDevPixels(
         [touch locationInView:self], [self contentScaleFactor]);
+#if defined(MOZ_IOS_LEGACY)
+    // majorRadius is iOS 8: assume a fingertip, 22pt across.
+    LayoutDeviceIntPoint radius =
+        UIKitPointsToDevPixels(CGPointMake(11, 11), [self contentScaleFactor]);
+#else
     LayoutDeviceIntPoint radius = UIKitPointsToDevPixels(
         CGPointMake([touch majorRadius], [touch majorRadius]),
         [self contentScaleFactor]);
+#endif
     NSNumber* value = [mTouches objectForKey:touch];
     if (value == nil) {
       // This shouldn't happen.
@@ -726,6 +734,7 @@ class nsAutoRetainUIKitObject {
   return [[self accessible] accessibilityElements];
 }
 
+#  if !defined(MOZ_IOS_LEGACY)
 - (UIAccessibilityContainerType)accessibilityContainerType {
   if (!mozilla::a11y::ShouldA11yBeEnabled()) {
     return [super accessibilityContainerType];
@@ -733,6 +742,7 @@ class nsAutoRetainUIKitObject {
 
   return [[self accessible] accessibilityContainerType];
 }
+#  endif
 #endif
 
 @end
@@ -1104,7 +1114,7 @@ void* nsWindow::GetNativeData(uint32_t aDataType) {
   return retVal;
 }
 
-CGFloat nsWindow::BackingScaleFactor() {
+double nsWindow::BackingScaleFactor() {
   if (mNativeView) {
     return [mNativeView contentScaleFactor];
   }
-- 
2.53.0

